Client/SafeBrowsing/GSafeBrowsing.swift

```diff
@@ -0,0 +1,411 @@
+                    } catch {
+                        return completion(nil, error)
+                    }
+                }
+            }
+            
+            do {
+                let response = try JSONDecoder().decode(type, from: data)
+                completion(response, nil)
+            } catch {
+                completion(nil, error)
+            }
+        }
+        task.resume()
+        return task
+    }
+    
+    private enum RequestType: String {
+        case get = "GET"
+        case post = "POST"
+    }
+    
+    private enum Endpoint: String {
+        case fetch = "/v4/threatListUpdates:fetch"
+        case fullHashes = "/v4/fullHashes:find"
+    }
+}
+
+extension URL {
+    func canonicalize() -> URL {
+        var absoluteString = self.absoluteString
+        absoluteString = absoluteString.replacingOccurrences(of: "\t", with: "")
+        absoluteString = absoluteString.replacingOccurrences(of: "\r", with: "")
+        absoluteString = absoluteString.replacingOccurrences(of: "\n", with: "")
+        
+        guard var components = URLComponents(string: absoluteString) else {
+            return self
+        }
+        
+        if var host = components.host?.removingPercentEncoding {
+            //TODO: Handle IP Addresses..
+            components.host = {
+                host = host.lowercased()
+                
+                while true {
+                    if host.hasPrefix(".") {
+                        host.removeFirst(1)
+                        continue
+                    }
+                    
+                    if host.hasSuffix(".") {
+                        host.removeLast(1)
+                        continue
+                    }
+                    
+                    break
+                }
+                return host
+            }()
+        }
+        
+        if var path = URL(string: absoluteString)?.pathComponents.map({ $0.removingPercentEncoding ?? $0 }) {
+            components.path = {
+                for i in 0..<path.count where path[i] == ".." {
+                    path[i] = ""
+                    path[i - 1] = ""
+                }
+                
+                return path.filter({ $0 != "." && !$0.isEmpty }).joined(separator: "/").replacingOccurrences(of: "//", with: "/")
+            }()
+        }
+        
+        if components.path.isEmpty {
+            components.path = "/"
+        }
+        
+        components.fragment = nil
+        components.port = nil
+        return components.url ?? self
+    }
+}
+
+extension URL {
+    private func calculatePrefixesAndSuffixes() -> [String] {
+        // Technically this should be done "TRIE" data structure
+        
+        //TODO: Fix for IP Address..
+        if let hostName = host?.replacingOccurrences(of: "\(scheme ?? "")://", with: "") {
+            var hostComponents = hostName.split(separator: ".")
+            while hostComponents.count > 5 {
```

Comment:
why 5 and why are we dropping excess components?

Reply:
The 5-component limit is in the spec: https://developers.google.com/safe-browsing/v4/urls-hashing#suffixprefix-expressions